docs/ProgrammaticApi.md

```diff
@@ -0,0 +1,125 @@
+---
+id: programmatic-api
+title: Programmatic API
+---
+
+:::caution
+
+The programmatic API is currently **experimental**. It is useful for advanced use cases only. You normally don't need to use it if you just want to run your tests.
+
+:::
+
+This page documents Jest's programmable API that can be used to run jest from `node`. TypeScript types are provided.
+
+## Simple example
+
+```js
+import {createJest} from 'jest';
+
+const jest = await createJest();
+jest.globalConfig = {
+  collectCoverage: false,
+  watch: false,
+  ...jest.globalConfig,
+};
```

Comment:
I don't think we should expose something that can be mutated in this way. I'd rather have a `mergeConfig` or some such that is invoked with the current config and expects to get a new config back it assigns internally

Reply:
Do you mean something like this?

```js
import {createJest, mergeConfig} from 'jest';
const jest = await createJest();

// Override global options
jest.globalConfig = mergeConfig(jest.globalConfig, {
  collectCoverage: false,
});
```

Or more like this?

```js
import {createJest} from 'jest';
const jest = await createJest();

// Override global options
jest.mergeConfig(jest.globalConfig, {
  collectCoverage: false,
});
```

Reply:
```js
import {createJest} from 'jest';
const jest = await createJest();

// Override global options
jest.mergeGlobalConfig({
  collectCoverage: false,
});
```

or

```js
jest.globalConfig.merge({
  collectCoverage: false,
});
```

The latter would mean `globalConfig` was some sort of class or something instead of just the object

Reply:
How would you override a project-specific config?

I think allowing you to mutate config in-place makes more sense, actually. So removing the read-only feature. I don't think freezing the config serves a real purpose. 🤷‍♀️

Reply:
I generally prefer encapsulation - makes it easier to debug when all mutations come though known methods and you can set a breakpoint in it.

---

>How would you override a project-specific config?

```js
jest.mergeProjectConfig(index, {
  testEnvironment: 'node',
});
```

This actually brings an important point to mind - any config modifications must go through its part in `normalize` within `jest-config`, to e.g. resolve `testEnvironment` to an absolute path.

I guess we can punt on that for now though. But if we _do_ kick it down the line for later, I want an

`jest.unstable_modifyConfig()` or some such so we can add proper config normalization later without a breaking change.